File: packages/sdk/src/bot/teamsBotSsoPrompt.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { AccessToken } from "@azure/identity";
import {
  Activity,
  ActivityTypes,
  CardFactory,
  Channels,
  MessageFactory,
  TurnContext,
  OAuthCard,
  ActionTypes,
  verifyStateOperationName,
  StatusCodes,
  TokenExchangeInvokeRequest,
  tokenExchangeOperationName,
  TokenExchangeResource,
  TeamsInfo,
  TeamsChannelAccount,
} from "botbuilder";
import {
  Dialog,
  DialogContext,
  DialogTurnResult,
  PromptOptions,
  PromptRecognizerResult,
} from "botbuilder-dialogs";
import { TeamsBotSsoPromptTokenResponse } from "./teamsBotSsoPromptTokenResponse";
import { v4 as uuidv4 } from "uuid";
import { ErrorWithCode, ErrorCode, ErrorMessage } from "../core/errors";
import { internalLogger } from "../util/logger";
import { validateScopesType, formatString, parseJwt, validateConfig } from "../util/utils";
import { OnBehalfOfCredentialAuthConfig } from "../models/configuration";
import { OnBehalfOfUserCredential } from "../credential/onBehalfOfUserCredential";

const invokeResponseType = "invokeResponse";
/**
 * Response body returned for a token exchange invoke activity.
 */
class TokenExchangeInvokeResponse {
  /**
   * Response id
   */
  id: string;

  /**
   * Detailed error message
   */
  failureDetail: string;

  constructor(id: string, failureDetail: string) {
    this.id = id;
    this.failureDetail = failureDetail;
  }
}

/**
 * Settings used to configure an TeamsBotSsoPrompt instance.
 */
export interface TeamsBotSsoPromptSettings {
  /**
   * The array of strings that declare the desired permissions and the resources requested.
   */
  scopes: string[];

  /**
   * (Optional) number of milliseconds the prompt will wait for the user to authenticate.
   * Defaults to a value `900,000` (15 minutes.)
   */
  timeout?: number;

  /**
   * (Optional) value indicating whether the TeamsBotSsoPrompt should end upon receiving an
   * invalid message.  Generally the TeamsBotSsoPrompt will end the auth flow when receives user
   * message not related to the auth flow. Setting the flag to false ignores the user's message instead.
   * Defaults to value `true`
   */
  endOnInvalidMessage?: boolean;
}

/**
 * Creates a new prompt that leverage Teams Single Sign On (SSO) support for bot to automatically sign in user and
 * help receive oauth token, asks the user to consent if needed.
 *
 * @remarks
 * The prompt will attempt to retrieve the users current token of the desired scopes and store it in
 * the token store.
 *
 * User will be automatically signed in leveraging Teams support of Bot Single Sign On(SSO):
 * https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/authentication/auth-aad-sso-bots
 *
 * @example
 * When used with your bots `DialogSet` you can simply add a new instance of the prompt as a named
 * dialog using `DialogSet.add()`. You can then start the prompt from a waterfall step using either
 * `DialogContext.beginDialog()` or `DialogContext.prompt()`. The user will be prompted to sign in as
 * needed and their access token will be passed as an argument to the callers next waterfall step:
 *
 * ```JavaScript
 * const { ConversationState, MemoryStorage } = require('botbuilder');
 * const { DialogSet, WaterfallDialog } = require('botbuilder-dialogs');
 * const { TeamsBotSsoPrompt } = require('@microsoft/teamsfx');
 *
 * const convoState = new ConversationState(new MemoryStorage());
 * const dialogState = convoState.createProperty('dialogState');
 * const dialogs = new DialogSet(dialogState);
 *
 * dialogs.add(new TeamsBotSsoPrompt('TeamsBotSsoPrompt', {
 *    scopes: ["User.Read"],
 * }));
 *
 * dialogs.add(new WaterfallDialog('taskNeedingLogin', [
 *      async (step) => {
 *          return await step.beginDialog('TeamsBotSsoPrompt');
 *      },
 *      async (step) => {
 *          const token = step.result;
 *          if (token) {
 *
 *              // ... continue with task needing access token ...
 *
 *          } else {
 *              await step.context.sendActivity(`Sorry... We couldn't log you in. Try again later.`);
 *              return await step.endDialog();
 *          }
 *      }
 * ]));
 * ```
 */
export class TeamsBotSsoPrompt extends Dialog {
  private authConfig: OnBehalfOfCredentialAuthConfig;
  private initiateLoginEndpoint: string;
  private settings: TeamsBotSsoPromptSettings;
  /**
   * Constructor of TeamsBotSsoPrompt.
   *
   * @param {OnBehalfOfCredentialAuthConfig} authConfig - Used to provide configuration and auth
   * @param {string} initiateLoginEndpoint - Login URL for Teams to redirect to
   * @param {string} dialogId Unique ID of the dialog within its parent `DialogSet` or `ComponentDialog`.
   * @param {TeamsBotSsoPromptSettings} settings Settings used to configure the prompt.
   *
   * @throws {@link ErrorCode|InvalidParameter} when scopes is not a valid string or string array.
   * @throws {@link ErrorCode|RuntimeNotSupported} when runtime is browser.
   */
  constructor(
    authConfig: OnBehalfOfCredentialAuthConfig,
    initiateLoginEndpoint: string,
    dialogId: string,
    settings: TeamsBotSsoPromptSettings
  ) {
    super(dialogId);

    this.initiateLoginEndpoint = initiateLoginEndpoint;
    this.authConfig = authConfig;
    this.settings = settings;

    validateScopesType(this.settings.scopes);
    validateConfig(this.authConfig);

    internalLogger.info("Create a new Teams Bot SSO Prompt");
  }

  /**
   * Called when a prompt dialog is pushed onto the dialog stack and is being activated.
   * @remarks
   * If the task is successful, the result indicates whether the prompt is still
   * active after the turn has been processed by the prompt.
   *
   * @param dc The DialogContext for the current turn of the conversation.
   *
   * @throws {@link ErrorCode|InvalidParameter} when timeout property in teams bot sso prompt settings is not number or is not positive.
   * @throws {@link ErrorCode|ChannelNotSupported} when bot channel is not MS Teams.
   * @throws {@link ErrorCode|RuntimeNotSupported} when runtime is browser.
   *
   * @returns A `Promise` representing the asynchronous operation.
   */
  public async beginDialog(dc: DialogContext): Promise<DialogTurnResult> {
    internalLogger.info("Begin Teams Bot SSO Prompt");
    this.ensureMsTeamsChannel(dc);

    // Initialize prompt state
    const default_timeout = 900000;
    let timeout: number = default_timeout;
    if (this.settings.timeout) {
      if (typeof this.settings.timeout != "number") {
        const errorMsg = "type of timeout property in teamsBotSsoPromptSettings should be number.";
        internalLogger.error(errorMsg);
        throw new ErrorWithCode(errorMsg, ErrorCode.InvalidParameter);
      }
      if (this.settings.timeout <= 0) {
        const errorMsg =
          "value of timeout property in teamsBotSsoPromptSettings should be positive.";
        internalLogger.error(errorMsg);
        throw new ErrorWithCode(errorMsg, ErrorCode.InvalidParameter);
      }
      timeout = this.settings.timeout;
    }

    if (this.settings.endOnInvalidMessage === undefined) {
      this.settings.endOnInvalidMessage = true;
    }
    const state: teamsBotSsoPromptState = dc.activeDialog?.state as teamsBotSsoPromptState;
    state.state = {};
    state.options = {};
    state.expires = new Date().getTime() + timeout;

    // Send OAuth card to get SSO token
    await this.sendOAuthCardAsync(dc.context);
    return Dialog.EndOfTurn;
  }

  /**
   * Called when a prompt dialog is the active dialog and the user replied with a new activity.
   *
   * @remarks
   * If the task is successful, the result indicates whether the dialog is still
   * active after the turn has been processed by the dialog.
   * The prompt generally continues to receive the user's replies until it accepts the
   * user's reply as valid input for the prompt.
   *
   * @param dc The DialogContext for the current turn of the conversation.
   *
   * @returns A `Promise` representing the asynchronous operation.
   *
   * @throws {@link ErrorCode|ChannelNotSupported} when bot channel is not MS Teams.
   * @throws {@link ErrorCode|RuntimeNotSupported} when runtime is browser.
   */
  public async continueDialog(dc: DialogContext): Promise<DialogTurnResult> {
    internalLogger.info("Continue Teams Bot SSO Prompt");
    this.ensureMsTeamsChannel(dc);

    // Check for timeout
    const state: teamsBotSsoPromptState = dc.activeDialog?.state as teamsBotSsoPromptState;
    const isMessage: boolean = dc.context.activity.type === ActivityTypes.Message;
    const isTimeoutActivityType: boolean =
      isMessage ||
      this.isTeamsVerificationInvoke(dc.context) ||
      this.isTokenExchangeRequestInvoke(dc.context);

    // If the incoming Activity is a message, or an Activity Type normally handled by TeamsBotSsoPrompt,
    // check to see if this TeamsBotSsoPrompt Expiration has elapsed, and end the dialog if so.
    const hasTimedOut: boolean = isTimeoutActivityType && new Date().getTime() > state.expires;
    if (hasTimedOut) {
      internalLogger.warn("End Teams Bot SSO Prompt due to timeout");
      return await dc.endDialog(undefined);
    } else {
      if (
        this.isTeamsVerificationInvoke(dc.context) ||
        this.isTokenExchangeRequestInvoke(dc.context)
      ) {
        // Recognize token
        const recognized: PromptRecognizerResult<TeamsBotSsoPromptTokenResponse> =
          await this.recognizeToken(dc);

        if (recognized.succeeded) {
          return await dc.endDialog(recognized.value);
        }
      } else if (isMessage && this.settings.endOnInvalidMessage) {
        internalLogger.warn("End Teams Bot SSO Prompt due to invalid message");
        return await dc.endDialog(undefined);
      }

      return Dialog.EndOfTurn;
    }
  }

  /**
   * Ensure bot is running in MS Teams since TeamsBotSsoPrompt is only supported in MS Teams channel.
   * @param dc dialog context
   * @throws {@link ErrorCode|ChannelNotSupported} if bot channel is not MS Teams
   * @internal
   */
  private ensureMsTeamsChannel(dc: DialogContext) {
    if (dc.context.activity.channelId != Channels.Msteams) {
      const errorMsg = formatString(
        ErrorMessage.OnlyMSTeamsChannelSupported,
        "Teams Bot SSO Prompt"
      );
      internalLogger.error(errorMsg);
      throw new ErrorWithCode(errorMsg, ErrorCode.ChannelNotSupported);
    }
  }

  /**
   * Send OAuthCard that tells Teams to obtain an authentication token for the bot application.
   * For details see https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/authentication/auth-aad-sso-bots.
   *
   * @internal
   */
  private async sendOAuthCardAsync(context: TurnContext): Promise<void> {
    internalLogger.verbose("Send OAuth card to get SSO token");

    const account: TeamsChannelAccount = await TeamsInfo.getMember(
      context,
      context.activity.from.id
    );
    internalLogger.verbose(
      "Get Teams member account user principal name: " +
        (account.userPrincipalName ? account.userPrincipalName : "")
    );

    const loginHint: string = account.userPrincipalName ? account.userPrincipalName : "";
    const signInResource = this.getSignInResource(loginHint);
    const card = CardFactory.oauthCard(
      "",
      "Teams SSO Sign In",
      "Sign In",
      signInResource.signInLink,
      signInResource.tokenExchangeResource
    );
    (card.content as OAuthCard).buttons[0].type = ActionTypes.Signin;
    const msg: Partial<Activity> = MessageFactory.attachment(card);

    // Send prompt
    await context.sendActivity(msg);
  }

  /**
   * Get sign in resource.
   *
   * @throws {@link ErrorCode|InvalidConfiguration} if client id, tenant id or initiate login endpoint is not found in config.
   *
   * @internal
   */
  private getSignInResource(loginHint: string) {
    internalLogger.verbose("Get sign in authentication configuration");

    const signInLink = `${this.initiateLoginEndpoint}?scope=${encodeURI(
      this.settings.scopes.join(" ")
    )}&clientId=${this.authConfig.clientId}&tenantId=${
      this.authConfig.tenantId
    }&loginHint=${loginHint}`;

    internalLogger.verbose("Sign in link: " + signInLink);

    const tokenExchangeResource: TokenExchangeResource = {
      id: uuidv4(),
    };

    return {
      signInLink: signInLink,
      tokenExchangeResource: tokenExchangeResource,
    };
  }

  /**
   * @internal
   */
  private async recognizeToken(
    dc: DialogContext
  ): Promise<PromptRecognizerResult<TeamsBotSsoPromptTokenResponse>> {
    const context = dc.context;
    let tokenResponse: TeamsBotSsoPromptTokenResponse | undefined;

    if (this.isTokenExchangeRequestInvoke(context)) {
      internalLogger.verbose("Receive token exchange request");
      // Received activity is not a token exchange request
      if (!(context.activity.value && this.isTokenExchangeRequest(context.activity.value))) {
        const warningMsg =
          "The bot received an InvokeActivity that is missing a TokenExchangeInvokeRequest value. This is required to be sent with the InvokeActivity.";

        internalLogger.warn(warningMsg);
        await context.sendActivity(
          this.getTokenExchangeInvokeResponse(StatusCodes.BAD_REQUEST, warningMsg)
        );
      } else {
        const ssoToken = context.activity.value.token;
        const credential = new OnBehalfOfUserCredential(ssoToken, this.authConfig);
        let exchangedToken: AccessToken | null;
        try {
          exchangedToken = await credential.getToken(this.settings.scopes);

          if (exchangedToken) {
            await context.sendActivity(
              this.getTokenExchangeInvokeResponse(StatusCodes.OK, "", context.activity.value.id)
            );

            const ssoTokenExpiration = parseJwt(ssoToken).exp;
            tokenResponse = {
              ssoToken: ssoToken,
              ssoTokenExpiration: new Date(ssoTokenExpiration * 1000).toISOString(),
              connectionName: "",
              token: exchangedToken.token,
              expiration: exchangedToken.expiresOnTimestamp.toString(),
            };
          }
        } catch (error) {
          const warningMsg = "The bot is unable to exchange token. Ask for user consent.";
          internalLogger.info(warningMsg);
          await context.sendActivity(
            this.getTokenExchangeInvokeResponse(
              StatusCodes.PRECONDITION_FAILED,
              warningMsg,
              context.activity.value.id
            )
          );
        }
      }
    } else if (this.isTeamsVerificationInvoke(context)) {
      internalLogger.verbose("Receive Teams state verification request");
      await this.sendOAuthCardAsync(dc.context);
      await context.sendActivity({ type: invokeResponseType, value: { status: StatusCodes.OK } });
    }

    return tokenResponse !== undefined
      ? { succeeded: true, value: tokenResponse }
      : { succeeded: false };
  }

  /**
   * @internal
   */
  private getTokenExchangeInvokeResponse(
    status: number,
    failureDetail: string,
    id?: string
  ): Activity {
    const invokeResponse: Partial<Activity> = {
      type: invokeResponseType,
      value: { status, body: new TokenExchangeInvokeResponse(id as string, failureDetail) },
    };
    return invokeResponse as Activity;
  }

  /**
   * @internal
   */
  private isTeamsVerificationInvoke(context: TurnContext): boolean {
    const activity: Activity = context.activity;

    return activity.type === ActivityTypes.Invoke && activity.name === verifyStateOperationName;
  }

  /**
   * @internal
   */
  private isTokenExchangeRequestInvoke(context: TurnContext): boolean {
    const activity: Activity = context.activity;

    return activity.type === ActivityTypes.Invoke && activity.name === tokenExchangeOperationName;
  }

  /**
   * @internal
   */
  private isTokenExchangeRequest(obj: any): obj is TokenExchangeInvokeRequest {
    return obj.hasOwnProperty("token");
  }
}

/**
 * @internal
 */
interface teamsBotSsoPromptState {
  state: any;
  options: PromptOptions;
  expires: number; // Timestamp of when the prompt will timeout.
}

File: packages/dotnet-sdk/src/TeamsFx/Bot/TeamsBotSsoPrompt.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema.Teams;
using Microsoft.Bot.Schema;
using System.Text.RegularExpressions;
using System.Net;
using Newtonsoft.Json.Linq;
using Microsoft.Bot.Connector;
using Microsoft.TeamsFx.Helper;
using Microsoft.Identity.Client;
using Microsoft.TeamsFx.Model;

namespace Microsoft.TeamsFx.Bot;

/// <summary>
/// Creates a new prompt that leverage Teams Single Sign On (SSO) support for bot to automatically sign in user and
/// help receive oauth token, asks the user to consent if needed.
/// </summary>
/// <remarks>
/// The prompt will attempt to retrieve the user's current token of the desired scopes.
/// User will be automatically signed in leveraging Teams support of Bot Single Sign On(SSO):
/// https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/authentication/auth-aad-sso-bots
/// </remarks>
/// 
/// <example>
/// ## Prompt Usage
///
/// When used with your bot's <see cref="DialogSet"/> you can simply add a new instance of the prompt as a named
/// dialog using <see cref="DialogSet.Add(Dialog)"/>. You can then start the prompt from a waterfall step using either
/// <see cref="DialogContext.BeginDialogAsync(string, object, CancellationToken)"/> or
/// <see cref="DialogContext.PromptAsync(string, PromptOptions, CancellationToken)"/>. The user
/// will be prompted to signin as needed and their access token will be passed as an argument to
/// the caller's next waterfall step.
/// 
/// <code>
/// var convoState = new ConversationState(new MemoryStorage());
/// var dialogState = convoState.CreateProperty&lt;DialogState&gt;("dialogState");
/// var dialogs = new DialogSet(dialogState); 
/// var botAuthOptions = new BotAuthenticationOptions { 
///     ClientId = "{client_id_guid_value}", 
///     ClientSecret = "{client_secret_value}", 
///     TenantId = "{tenant_id_guid_value}", 
///     ApplicationIdUri = "{application_id_uri_value}", 
///     OAuthAuthority = "https://login.microsoftonline.com/{tenant_id_guid_value}", 
///     LoginStartPageEndpoint = "https://{bot_web_app_domain}/bot-auth-start" 
///     };
///     
/// var scopes = new string[] { "User.Read" };
/// var teamsBotSsoPromptSettings = new TeamsBotSsoPromptSettings(botAuthOptions, scopes);
/// 
/// dialogs.Add(new TeamsBotSsoPrompt("{unique_id_for_the_prompt}", teamsBotSsoPromptSettings)); 
/// dialogs.Add(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] 
/// { 
///     async(WaterfallStepContext stepContext, CancellationToken cancellationToken) => {
///         return await stepContext.BeginDialogAsync(nameof(TeamsBotSsoPrompt), null, cancellationToken);
///     }, 
///     async(WaterfallStepContext stepContext, CancellationToken cancellationToken) => { 
///         var tokenResponse = (TeamsBotSsoPromptTokenResponse)stepContext.Result; 
///         if (tokenResponse?.Token != null) 
///         { 
///             // ... continue with task needing access token ... 
///         } 
///         else
///         {
///             await stepContext.Context.SendActivityAsync(MessageFactory.Text("Login was not successful please try again."), cancellationToken);
///         }
///         return await stepContext.EndDialogAsync(cancellationToken: cancellationToken); 
///     } 
/// }));
/// </code>
/// 
/// </example>
public class TeamsBotSsoPrompt : Dialog
{
    private readonly TeamsBotSsoPromptSettings _settings;
    private const string PersistedExpires = "expires";
    internal IIdentityClientAdapter _identityClientAdapter { private get; set; }
    internal ITeamsInfo _teamsInfo { private get; set; }


    /// <summary>
    /// Initializes a new instance of the <see cref="TeamsBotSsoPrompt"/> class.
    /// </summary>
    /// <param name="dialogId">The ID to assign to this prompt.</param>
    /// <param name="settings">Additional OAuth settings to use with this instance of the prompt.</param>
    /// <remarks>The value of <paramref name="dialogId"/> must be unique within the
    /// <see cref="DialogSet"/> or <see cref="ComponentDialog"/> to which the prompt is added.</remarks>
    /// <exception cref="ExceptionCode.InvalidParameter">When input parameters is null.</exception>
    public TeamsBotSsoPrompt(string dialogId, TeamsBotSsoPromptSettings settings): base(dialogId)
    {
        if (string.IsNullOrWhiteSpace(dialogId))
        {
            throw new ExceptionWithCode($"Parameter {nameof(dialogId)} is null or empty.", ExceptionCode.InvalidParameter);
        }
        _settings = settings ?? throw new ExceptionWithCode($"Parameter {nameof(settings)} is null or empty.", ExceptionCode.InvalidParameter);

        var confidentialClientApplication = ConfidentialClientApplicationBuilder.Create(_settings.BotAuthOptions.ClientId)
            .WithClientSecret(_settings.BotAuthOptions.ClientSecret)
            .WithAuthority(_settings.BotAuthOptions.OAuthAuthority)
            .Build();
        _identityClientAdapter = new IdentityClientAdapter(confidentialClientApplication);
        _teamsInfo = new TeamsInfoWrapper();
    }

    /// <summary>
    /// Called when the dialog is started and pushed onto the dialog stack.
    /// </summary>
    /// <param name="dialogContext">The Microsoft.Bot.Builder.Dialogs.DialogContext for the current turn of conversation.</param>
    /// <param name="options">Optional, initial information to pass to the dialog.</param>
    /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
    /// <returns> A System.Threading.Tasks.Task representing the asynchronous operation.</returns>
    /// <exception cref="ExceptionCode.InvalidParameter">if dialog context argument is null</exception>
    public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dialogContext, object options = null, CancellationToken cancellationToken = default)
    {
        if (dialogContext == null)
        {
            throw new ExceptionWithCode($"Parameter {nameof(dialogContext)} is null or empty.", ExceptionCode.InvalidParameter);
        }

        EnsureMsTeamsChannel(dialogContext);

        var state = dialogContext.ActiveDialog?.State;
        state[PersistedExpires] = DateTime.UtcNow.AddMilliseconds(_settings.Timeout);

        // Send OAuthCard that tells Teams to obtain an authentication token for the bot application.
        await SendOAuthCardToObtainTokenAsync(dialogContext.Context, cancellationToken).ConfigureAwait(false);
        return EndOfTurn;
    }

    /// <summary>
    /// Called when a prompt dialog is the active dialog and the user replied with a new activity.
    /// </summary>
    /// <param name="dc">The <see cref="DialogContext"/> for the current turn of conversation.</param>
    /// <param name="cancellationToken">A cancellation token that can be used by other objects
    /// or threads to receive notice of cancellation.</param>
    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
    /// <remarks>If the task is successful, the result indicates whether the dialog is still
    /// active after the turn has been processed by the dialog.
    /// <para>The prompt generally ends on invalid message from user's reply.</para></remarks>
    /// <exception cref="ExceptionCode.InternalError">When failed to login with unknown error.</exception>
    /// <exception cref="ExceptionCode.ServiceError">When failed to get access token from identity server(AAD).</exception>
    public override async Task<DialogTurnResult> ContinueDialogAsync(DialogContext dc, CancellationToken cancellationToken = default(CancellationToken))
    {
        EnsureMsTeamsChannel(dc);

        // Check for timeout
        var state = dc.ActiveDialog?.State;
        bool isMessage = (dc.Context.Activity.Type == ActivityTypes.Message);
        bool isTimeoutActivityType =
          isMessage ||
          IsTeamsVerificationInvoke(dc.Context) ||
          IsTokenExchangeRequestInvoke(dc.Context);

        // If the incoming Activity is a message, or an Activity Type normally handled by TeamsBotSsoPrompt,
        // check to see if this TeamsBotSsoPrompt Expiration has elapsed, and end the dialog if so.
        bool hasTimedOut = isTimeoutActivityType && DateTime.Compare(DateTime.UtcNow, (DateTime)state[PersistedExpires]) > 0;
        if (hasTimedOut)
        {
            return await dc.EndDialogAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
        }
        else
        {
            if (IsTeamsVerificationInvoke(dc.Context) || IsTokenExchangeRequestInvoke(dc.Context))
            {
                // Recognize token
                PromptRecognizerResult<TeamsBotSsoPromptTokenResponse> recognized = await RecognizeTokenAsync(dc, cancellationToken).ConfigureAwait(false);

                if (recognized.Succeeded)
                {
                    return await dc.EndDialogAsync(recognized.Value, cancellationToken).ConfigureAwait(false);
                }
            }
            else if (isMessage)
            {
                return await dc.EndDialogAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
            }

            return EndOfTurn;
        }
    }

    /// <summary>
    /// This is intended for internal use.
    /// </summary>
    /// <param name="dc">DialogContext.</param>
    /// <param name="cancellationToken">CancellationToken.</param>
    /// <returns>PromptRecognizerResult.</returns>
    /// <exception cref="ExceptionCode.InternalError">When failed to login with unknown error.</exception>
    /// <exception cref="ExceptionCode.ServiceError">When failed to get access token from identity server(AAD).</exception>
    private async Task<PromptRecognizerResult<TeamsBotSsoPromptTokenResponse>> RecognizeTokenAsync(DialogContext dc, CancellationToken cancellationToken)
    {

        ITurnContext context = dc.Context;
        var result = new PromptRecognizerResult<TeamsBotSsoPromptTokenResponse>();
        TeamsBotSsoPromptTokenResponse tokenResponse = null;

        if (IsTokenExchangeRequestInvoke(context))
        {
            var tokenResponseObject = context.Activity.Value as JObject;
            string ssoToken = tokenResponseObject?.ToObject<TokenExchangeInvokeRequest>().Token;
            // Received activity is not a token exchange request
            if (String.IsNullOrEmpty(ssoToken))
            {
                var warningMsg =
                  "The bot received an InvokeActivity that is missing a TokenExchangeInvokeRequest value. This is required to be sent with the InvokeActivity.";
                await SendInvokeResponseAsync(context, HttpStatusCode.BadRequest, warningMsg, cancellationToken).ConfigureAwait(false);
            }
            else
            {
                try
                {
                    var exchangedToken = await GetToken(ssoToken, _settings.Scopes).ConfigureAwait(false);

                    var ssoTokenObj = Utils.ParseJwt(ssoToken);
                    var ssoExpiration = DateTimeOffset.FromUnixTimeSeconds(long.Parse(ssoTokenObj.Payload["exp"].ToString()));
                    tokenResponse = new TeamsBotSsoPromptTokenResponse {
                        SsoToken = ssoToken,
                        SsoTokenExpiration = ssoExpiration.ToString(),
                        Token = exchangedToken.Token,
                        Expiration = exchangedToken.ExpiresOn.ToString()
                    };

                    await SendInvokeResponseAsync(context, HttpStatusCode.OK, null, cancellationToken).ConfigureAwait(false);
                }
                catch (MsalUiRequiredException) // Need user interaction
                {
                    var warningMsg = "The bot is unable to exchange token. Ask for user consent first.";
                    await SendInvokeResponseAsync(context, HttpStatusCode.PreconditionFailed, new TokenExchangeInvokeResponse {
                        Id = context.Activity.Id,
                        FailureDetail = warningMsg,
                    }, cancellationToken).ConfigureAwait(false);
                }
                catch (MsalServiceException ex) // Errors that returned from AAD service
                {
                    throw new ExceptionWithCode($"Failed to get access token from OAuth identity server with error: {ex.ResponseBody}", ExceptionCode.ServiceError);
                }
                catch (MsalClientException ex) // Exceptions that are local to the MSAL library
                {
                    throw new ExceptionWithCode($"Failed to get access token with error: {ex.Message}", ExceptionCode.InternalError);
                }

            }
        }
        else if (IsTeamsVerificationInvoke(context))
        {
            await SendOAuthCardToObtainTokenAsync(context, cancellationToken).ConfigureAwait(false);
            await SendInvokeResponseAsync(context, HttpStatusCode.OK, null, cancellationToken).ConfigureAwait(false);
        }

        if (tokenResponse != null)
        {
            result.Succeeded = true;
            result.Value = tokenResponse;
        } else
        {
            result.Succeeded = false;
        }
        return result;
    }

    private async Task<AccessToken> GetToken(string ssoToken, string[] scopes)
    {
        AccessToken result;
        var ssoTokenObj = Utils.ParseJwt(ssoToken);
        var ssoTokenExpiration = DateTimeOffset.FromUnixTimeSeconds(long.Parse(ssoTokenObj.Payload["exp"].ToString())); 
        
        // Get sso token
        if (scopes.Length == 0)
        {
            if (DateTimeOffset.Compare(DateTimeOffset.UtcNow, ssoTokenExpiration) > 0)
            {
                throw new ExceptionWithCode("SSO token has already expired.", ExceptionCode.TokenExpiredError);
            }
            result = new AccessToken(ssoToken, ssoTokenExpiration);
        }
        else
        {
            var authenticationResult = await _identityClientAdapter.GetAccessToken(ssoToken, scopes).ConfigureAwait(false);
            result = new AccessToken(authenticationResult.AccessToken, authenticationResult.ExpiresOn);
        }
        return result;
    }

    private static async Task SendInvokeResponseAsync(ITurnContext turnContext, HttpStatusCode statusCode, object body, CancellationToken cancellationToken)
    {
        await turnContext.SendActivityAsync(
            new Activity {
                Type = ActivityTypesEx.InvokeResponse,
                Value = new InvokeResponse {
                    Status = (int)statusCode,
                    Body = body,
                },
            }, cancellationToken).ConfigureAwait(false);
    }

    private bool IsTeamsVerificationInvoke(ITurnContext context) {
       return (context.Activity.Type == ActivityTypes.Invoke) && (context.Activity.Name == SignInConstants.VerifyStateOperationName);
    }
    private bool IsTokenExchangeRequestInvoke(ITurnContext context) {
        return (context.Activity.Type == ActivityTypes.Invoke) && (context.Activity.Name == SignInConstants.TokenExchangeOperationName);
    }

    /// <summary>
    /// Send OAuthCard that tells Teams to obtain an authentication token for the bot application.
    /// For details see https://docs.microsoft.com/en-us/microsoftteams/platform/bots/how-to/authentication/auth-aad-sso-bots.
    /// </summary>
    /// <param name="context">ITurnContext</param>
    /// <param name="cancellationToken">CancellationToken.</param>
    /// <returns>The task to await.</returns>
    private async Task SendOAuthCardToObtainTokenAsync(ITurnContext context, CancellationToken cancellationToken)
    {
        TeamsChannelAccount account = await _teamsInfo.GetTeamsMemberAsync(context, context.Activity.From.Id, cancellationToken).ConfigureAwait(false);

        string loginHint = account.UserPrincipalName ?? "";
        if (String.IsNullOrEmpty(account.TenantId))
        {
            throw new ExceptionWithCode("Failed to get tenant id through bot framework.", ExceptionCode.ServiceError);
        }
        string tenantId = account.TenantId ?? "";
        SignInResource signInResource = GetSignInResource(loginHint, tenantId);

        // Ensure prompt initialized
        IMessageActivity prompt = Activity.CreateMessageActivity();
        prompt.Attachments = new List<Attachment>();
        prompt.Attachments.Add(new Attachment {
            ContentType = OAuthCard.ContentType,
            Content = new OAuthCard {
                Text = "Sign In",
                Buttons = new[]
                {
                            new CardAction
                            {
                                    Title = "Teams SSO Sign In",
                                    Value = signInResource.SignInLink,
                                    Type = ActionTypes.Signin,
                            },
                        },
                TokenExchangeResource = signInResource.TokenExchangeResource,
            },
        });
        // Send prompt
        await context.SendActivityAsync(prompt, cancellationToken).ConfigureAwait(false);
    }


    /// <summary>
    /// Get sign in authentication configuration
    /// </summary>
    /// <param name="loginHint">login hint</param>
    /// <param name="tenantId">tenant id</param>
    /// <returns>sign in resource</returns>
    private SignInResource GetSignInResource(string loginHint, string tenantId)
    {
        string signInLink = $"{_settings.BotAuthOptions.InitiateLoginEndpoint }?scope={Uri.EscapeDataString(string.Join(" ", _settings.Scopes))}&clientId={_settings.BotAuthOptions.ClientId}&tenantId={tenantId}&loginHint={loginHint}";

        SignInResource signInResource = new SignInResource {
            SignInLink = signInLink,
            TokenExchangeResource = new TokenExchangeResource {
                Id = Guid.NewGuid().ToString(),
                Uri = Regex.Replace(_settings.BotAuthOptions.ApplicationIdUri, @"/\/$/", "") + "/access_as_user"
            }
        };

        return signInResource;
    }

    /// <summary>
    /// Ensure bot is running in MS Teams since TeamsBotSsoPrompt is only supported in MS Teams channel.
    /// </summary>
    /// <param name="dc">dialog context</param>
    /// <exception cref="ExceptionCode.ChannelNotSupported"> if bot channel is not MS Teams </exception>
    private void EnsureMsTeamsChannel(DialogContext dc)
    {
        if (dc.Context.Activity.ChannelId != Channels.Msteams)
        {
            var errorMessage = "Teams Bot SSO Prompt is only supported in MS Teams Channel";
            throw new ExceptionWithCode(errorMessage, ExceptionCode.ChannelNotSupported);
        }
    }
}
